plugins/hls-cabal-plugin/src/Ide/Plugin/Cabal/Completions.hs

```diff
@@ -0,0 +1,656 @@
+  }
+  deriving (Eq, Show)
+
+-- ----------------------------------------------------------------
+-- Public API for Completions
+-- ----------------------------------------------------------------
+
+{- | Takes information about the completion status within the file
+  and finds the correct completer to be applied
+-}
+contextToCompleter :: Context -> Completer
+-- if we are in the top level of the cabal file and not in a keyword context,
+-- we can write any top level keywords or a stanza declaration
+contextToCompleter (TopLevel, None) =
+  constantCompleter $
+    Map.keys (cabalVersionKeyword <> cabalKeywords) ++ Map.keys stanzaKeywordMap
+-- if we are in a keyword context in the top level,
+-- we look up that keyword in the top level context and can complete its possible values
+contextToCompleter (TopLevel, KeyWord kw) =
+  case Map.lookup kw (cabalVersionKeyword <> cabalKeywords) of
+    Nothing -> noopCompleter
+    Just l  -> l
+-- if we are in a stanza and not in a keyword context,
+-- we can write any of the stanza's keywords or a stanza declaration
+contextToCompleter (Stanza s, None) =
+  case Map.lookup s stanzaKeywordMap of
+    Nothing -> noopCompleter
+    Just l  -> constantCompleter $ Map.keys l ++ Map.keys stanzaKeywordMap
+-- if we are in a stanza's keyword's context we can complete possible values of that keyword
+contextToCompleter (Stanza s, KeyWord kw) =
+  case Map.lookup s stanzaKeywordMap of
+    Nothing -> noopCompleter
+    Just m -> case Map.lookup kw m of
+      Nothing -> noopCompleter
+      Just l  -> l
+
+{- | Takes information about the current cursor position,
+  the handled cabal file and a set of possible keywords
+  and creates completion suggestions that fit the current input
+  from the given list
+-}
+mkCompletionItems :: [CabalCompletionItem] -> [LSP.CompletionItem]
+mkCompletionItems l = map buildCompletion l
+
+{- | Takes a position and a list of lines (representing a file)
+  and returns the context of the current position
+  can return Nothing if an error occurs
+  TODO: first line can only have cabal-version: keyword
+-}
+getContext :: CabalCompletionContext -> [T.Text] -> Maybe Context
+getContext ctx ls =
+  case lvlContext of
+    TopLevel -> do
+      kwContext <- getKeyWordContext ctx ls (cabalVersionKeyword <> cabalKeywords)
+      pure (TopLevel, kwContext)
+    Stanza s ->
+      case Map.lookup s stanzaKeywordMap of
+        Nothing -> do
+          pure (Stanza s, None)
+        Just m -> do
+          kwContext <- getKeyWordContext ctx ls m
+          pure (Stanza s, kwContext)
+ where
+  pos = completionCursorPosition ctx
+  lvlContext =
+    if pos ^. JL.character == 0
+      then TopLevel
+      else currentLevel (previousLines pos ls)
+
+-- ----------------------------------------------------------------
+-- Helper Functions
+-- ----------------------------------------------------------------
+
+{- | Takes a position, a list of lines (representing a file)
```

Comment:
A few times you say "position" when I think you maybe want to say "context"?

Reply:
Very true, thank you!